File: src/main/java/spring/springserver/global/jwt/JwtAuthFilter.java
package spring.springserver.global.jwt;

import jakarta.servlet.FilterChain;
import jakarta.servlet.ServletException;
import jakarta.servlet.http.HttpServletRequest;
import jakarta.servlet.http.HttpServletResponse;
import lombok.RequiredArgsConstructor;
import org.jspecify.annotations.NonNull;
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken;
import org.springframework.security.core.authority.SimpleGrantedAuthority;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.stereotype.Component;
import org.springframework.web.filter.OncePerRequestFilter;
import spring.springserver.domain.auth.interfaces.TokenProvider;

import java.io.IOException;
import java.util.Collections;

@Component
@RequiredArgsConstructor
public class JwtAuthFilter extends OncePerRequestFilter {

	private final TokenProvider tokenProvider;

	@Override
	protected void doFilterInternal(@NonNull HttpServletRequest request,
									@NonNull HttpServletResponse response,
									@NonNull FilterChain filterChain) throws ServletException, IOException {

		String token = tokenProvider.resolveToken(request);

		if (token == null || tokenProvider.isValidToken(token)) {

			filterChain.doFilter(request, response);
			return;
		}

		String username = tokenProvider.getUsernameFromToken(token);
		String role = tokenProvider.getRole(token).name();

		SimpleGrantedAuthority authority = new SimpleGrantedAuthority(role.startsWith("ROLE_") ? role : "ROLE_" + role);

		UsernamePasswordAuthenticationToken authenticationToken = new UsernamePasswordAuthenticationToken(
				username,
				null,
				Collections.singletonList(authority)
		);

		SecurityContextHolder.getContext().setAuthentication(authenticationToken);
		filterChain.doFilter(request, response);
	}
}

File: src/main/java/spring/springserver/global/jwt/JwtAuthFilter.kt
package spring.springserver.global.jwt

import jakarta.servlet.FilterChain
import jakarta.servlet.http.HttpServletRequest
import jakarta.servlet.http.HttpServletResponse
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken
import org.springframework.security.core.context.SecurityContextHolder
import org.springframework.stereotype.Component
import org.springframework.web.filter.OncePerRequestFilter

@Component
class JwtAuthFilter(
    private val tokenProvider: TokenProvider,
    private val memberDetailsService: MemberDetailsService
): OncePerRequestFilter() {

    override fun doFilterInternal(
        httpServletRequest: HttpServletRequest,
        httpServletResponse: HttpServletResponse,
        filterChain: FilterChain
    ) {

        val token = tokenProvider.resolveToken(httpServletRequest)

        if (token == null || tokenProvider.isNotValidToken(token) || tokenProvider.isNotAccessToken(token)) {

            filterChain.doFilter(
                httpServletRequest,
                httpServletResponse
            )

            return
        }

        val username = tokenProvider.getUsernameFromToken(token)

        if (username.isNullOrBlank()) {

            filterChain.doFilter(
                httpServletRequest,
                httpServletResponse
            )

            return
        }

        val memberDetails = runCatching {

            memberDetailsService.loadUserByUsername(username = username)
        }.getOrNull() as? MemberDetails

        if (memberDetails == null) {

            filterChain.doFilter(
                httpServletRequest,
                httpServletResponse
            )

            return
        }

        val authenticationToken = UsernamePasswordAuthenticationToken(
            memberDetails,
            null,
            memberDetails.authorities
        )

        SecurityContextHolder.getContext().authentication = authenticationToken

        filterChain.doFilter(
            httpServletRequest,
            httpServletResponse
        )
    }
}
